### ISSUE TITLE
Deep equality matching broken on arrays

### ISSUE THREAD

**Issue by ethanresnick** [sinonjs/samsam#54]
* library version : 3.0.2
* Example URL   :
* Other libraries you are using: None

**What did you expect to happen?**
Returns true

**What actually happens**
Returns false

**How to reproduce**
```js
var samsam = require('@sinonjs/samsam');
samsam.match(
  [[""]], 
  samsam.match.array.deepEquals(
    [[""]]
  )
);
```

**Comment by mantoni** [comment]
Thank you for reporting. You're using `match` with a `matcher` function. You can simply do this:

```js
samsam.match([['x', 'test']], [['test']])
```

The `matcher` functions can be used with `deepEqual` like this:

```js
samsam.deepEqual([['test']], [samsam.match.array.deepEquals(['test'])])
```

I'm not 100% sure if samsam should support your use case – or maybe even throw a meaningful exception. What do you think @mroderick?

However, while running some local tests, I found that the following is not working as I would expect:

```js
samsam.deepEqual([['test']], samsam.match.array.deepEquals([['test']]))
```

So in any case, I think matching arrays needs a little improvement.

**Comment by mroderick** [comment]
I think that improving how arrays are matched would be a good idea. Does anyone have time to study this in more detail and write up a solution proposal for it?

**Comment by mantoni** [comment]
This was implemented in #55. While additional improvements to array deep equality checks might be coming, I think this issue is now resolved.
